File: ios/Mynah/UI/HomeCoach.swift
import SwiftUI

/// Where the first-run tour points. Measured on the composer in screen
/// coordinates and handed up, because the overlay is drawn above everything and
/// so cannot live inside the screen it is describing.
///
/// `.zero` means "not on screen" — the mic is absent on a phone that cannot
/// transcribe the interface language, and a tour step pointing at nothing is
/// worse than one fewer step.
struct CoachTargets: Equatable {
    var composer: CGRect = .zero
    var mic: CGRect = .zero
    var action: CGRect = .zero
}

private struct CoachStep {
    let target: CGRect
    let corner: CGFloat
    let title: String
    let body: String
}

/// The three coach marks over the composer: say it in one sentence, hold to
/// speak, nothing dials until you confirm.
///
/// A 62% ink scrim with the target punched out of it and ringed in lime. The
/// hole is cut with a blend mode rather than by painting four rectangles around
/// the target: that would leave seams at the corners, and cannot round them.
struct HomeCoach: View {
    let targets: CoachTargets
    let onFinished: () -> Void

    @State private var index = 0

    private var steps: [CoachStep] {
        var steps: [CoachStep] = []
        if !targets.composer.isEmpty {
            steps.append(CoachStep(target: targets.composer, corner: 30,
                                   title: t("coach_title_1"), body: t("coach_body_1")))
        }
        if !targets.mic.isEmpty {
            steps.append(CoachStep(target: targets.mic, corner: targets.mic.height / 2,
                                   title: t("coach_title_2"), body: t("coach_body_2")))
        }
        if !targets.action.isEmpty {
            steps.append(CoachStep(target: targets.action, corner: 34,
                                   title: t("coach_title_3"), body: t("coach_body_3")))
        }
        return steps
    }

    var body: some View {
        // Nothing measured yet: the first frame of the composer has no bounds
        // to point at.
        if steps.isEmpty {
            EmptyView()
        } else {
            let step = steps[min(index, steps.count - 1)]
            let hole = step.target.insetBy(dx: -6, dy: -6)
            let last = index >= steps.count - 1

            GeometryReader { geometry in
                ZStack(alignment: .topLeading) {
                    Ink.text.opacity(0.62)
                        .ignoresSafeArea()
                        .compositingGroup()
                        .overlay(alignment: .topLeading) {
                            RoundedRectangle(cornerRadius: step.corner, style: .continuous)
                                .frame(width: hole.width, height: hole.height)
                                .position(x: hole.midX, y: hole.midY)
                                .blendMode(.destinationOut)
                                .ignoresSafeArea()
                        }
                        .compositingGroup()

                    // The ring, drawn after the hole so the punch cannot erase it.
                    RoundedRectangle(cornerRadius: step.corner, style: .continuous)
                        .strokeBorder(Ink.lime, lineWidth: 2.5)
                        .frame(width: hole.width, height: hole.height)
                        .position(x: hole.midX, y: hole.midY)
                        .ignoresSafeArea()

                    CoachCard(
                        hole: hole,
                        available: geometry.size.height,
                        index: index,
                        total: steps.count,
                        title: step.title,
                        body: step.body,
                        last: last,
                        onSkip: onFinished,
                        onNext: { if last { onFinished() } else { index += 1 } }
                    )
                }
                // Swallows everything: during the tour the screen underneath is
                // an illustration, not something to poke at.
                .contentShape(Rectangle())
                .onTapGesture {}
            }
            .ignoresSafeArea()
        }
    }
}

private struct CoachCard: View {
    let hole: CGRect
    let available: CGFloat
    let index: Int
    let total: Int
    let title: String
    let body_: String
    let last: Bool
    let onSkip: () -> Void
    let onNext: () -> Void

    init(
        hole: CGRect, available: CGFloat, index: Int, total: Int,
        title: String, body: String, last: Bool,
        onSkip: @escaping () -> Void, onNext: @escaping () -> Void
    ) {
        self.hole = hole
        self.available = available
        self.index = index
        self.total = total
        self.title = title
        self.body_ = body
        self.last = last
        self.onSkip = onSkip
        self.onNext = onNext
    }

    @State private var height: CGFloat = 0

    private let margin: CGFloat = 24
    private let gap: CGFloat = 20

    /// Below the hole when there is room under it, above when there is not —
    /// the card must never cover what it points at.
    private var offset: CGFloat {
        let below = hole.maxY + gap
        let fitsBelow = below + height <= available - margin
        return max(margin, fitsBelow ? below : hole.minY - gap - height)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text("\(index + 1) / \(total)").wise(Type.labelSmall).foregroundStyle(Ink.mute)
            Spacer().frame(height: 6)
            Text(title).wise(Type.section).foregroundStyle(Ink.text)
            Spacer().frame(height: 6)
            Text(body_)
                .wise(Type.caption)
                .foregroundStyle(Ink.body)
                .fixedSize(horizontal: false, vertical: true)
            Spacer().frame(height: 14)
            HStack(spacing: 0) {
                // The way out is only offered while there is something left to
                // skip.
                if !last {
                    LinkText(t("action_skip_tour"), style: Type.linkSmall, action: onSkip)
                }
                Spacer(minLength: 12)
                PrimaryButton(
                    t(last ? "action_got_it" : "action_next"),
                    height: 40,
                    style: Type.buttonSmall,
                    action: onNext
                )
                .fixedSize()
            }
        }
        .padding(18)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Ink.card, in: cardShape)
        .background(
            GeometryReader { geometry in
                Color.clear.onAppear { height = geometry.size.height }
                    .onChange(of: geometry.size.height) { _, value in height = value }
            }
        )
        .padding(.horizontal, 24)
        .offset(y: offset)
    }
}

File: android/app/src/main/java/com/voicecall/ui/HomeCoach.kt
package com.voicecall.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.CompositingStrategy
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import androidx.compose.material3.Text
import com.voicecall.R
import com.voicecall.ui.theme.Ink
import com.voicecall.ui.theme.Type
import kotlin.math.roundToInt

/**
 * Where the first-run tour points. Measured on Home in window coordinates and
 * handed up, because the overlay is drawn above the navigation bar and so
 * cannot live inside the screen it is describing.
 *
 * [Rect.Zero] means "not on screen" — the mic is absent on a device with no
 * speech recogniser, and a tour step pointing at nothing is worse than one
 * fewer step.
 */
data class CoachTargets(
    val composer: Rect = Rect.Zero,
    val mic: Rect = Rect.Zero,
    val action: Rect = Rect.Zero,
)

private data class CoachStep(val target: Rect, val corner: Float, val title: Int, val body: Int)

/**
 * The three coach marks over Home: say it in one sentence, hold to speak,
 * nothing dials until you confirm.
 *
 * A 62% ink scrim with the target punched out of it and ringed in lime. The
 * hole is cut with a Clear blend inside an offscreen layer — painting four
 * rectangles around the target instead would leave seams on the corners, and
 * cannot round them.
 */
@Composable
fun HomeCoach(targets: CoachTargets, onFinished: () -> Unit) {
    val steps = remember(targets) {
        listOfNotNull(
            targets.composer.takeIf { !it.isEmpty }
                ?.let { CoachStep(it, 30f, R.string.coach_title_1, R.string.coach_body_1) },
            targets.mic.takeIf { !it.isEmpty }
                ?.let { CoachStep(it, it.height / 2f, R.string.coach_title_2, R.string.coach_body_2) },
            targets.action.takeIf { !it.isEmpty }
                ?.let { CoachStep(it, 34f, R.string.coach_title_3, R.string.coach_body_3) },
        )
    }

    var index by remember { mutableIntStateOf(0) }
    // Nothing measured yet: the first frame of Home has no bounds to point at.
    if (steps.isEmpty()) return
    val step = steps[index.coerceAtMost(steps.lastIndex)]
    val last = index >= steps.lastIndex

    val density = LocalDensity.current
    val pad = with(density) { 6.dp.toPx() }
    val hole = Rect(
        left = step.target.left - pad,
        top = step.target.top - pad,
        right = step.target.right + pad,
        bottom = step.target.bottom + pad,
    )

    Box(
        Modifier
            .fillMaxSize()
            // Swallows everything: during the tour the screen underneath is an
            // illustration, not something to poke at.
            .pointerInput(steps, index) { detectTapGestures { } }
            .graphicsLayer { compositingStrategy = CompositingStrategy.Offscreen }
            .drawWithContent {
                drawRect(Ink.Text.copy(alpha = 0.62f))
                drawRoundRect(
                    color = Color.Transparent,
                    topLeft = Offset(hole.left, hole.top),
                    size = Size(hole.width, hole.height),
                    cornerRadius = CornerRadius(step.corner * density.density),
                    blendMode = BlendMode.Clear,
                )
                drawContent()
            },
    ) {
        // The ring, drawn after the hole so the Clear pass cannot erase it.
        Box(
            Modifier.fillMaxSize().drawBehind {
                drawRoundRect(
                    color = Ink.Lime,
                    topLeft = Offset(hole.left, hole.top),
                    size = Size(hole.width, hole.height),
                    cornerRadius = CornerRadius(step.corner * density.density),
                    style = Stroke(width = 2.5.dp.toPx()),
                )
            },
        )

        CoachCard(
            hole = hole,
            index = index,
            total = steps.size,
            title = stringResource(step.title),
            body = stringResource(step.body),
            last = last,
            onSkip = onFinished,
            onNext = { if (last) onFinished() else index++ },
        )
    }
}

@Composable
private fun CoachCard(
    hole: Rect,
    index: Int,
    total: Int,
    title: String,
    body: String,
    last: Boolean,
    onSkip: () -> Unit,
    onNext: () -> Unit,
) {
    val density = LocalDensity.current
    var cardHeight by remember { mutableIntStateOf(0) }
    var containerHeight by remember { mutableIntStateOf(0) }

    val margin = with(density) { 24.dp.toPx() }
    val gap = with(density) { 20.dp.toPx() }

    Box(
        Modifier
            .fillMaxSize()
            .onSizeChanged { containerHeight = it.height }
            .padding(horizontal = 24.dp),
    ) {
        Box(
            Modifier
                .fillMaxWidth()
                .offset {
                    // Below the hole when there is room under it, above when
                    // there is not — the card must never cover what it points at.
                    val below = hole.bottom + gap
                    val fitsBelow = below + cardHeight <= containerHeight - margin
                    val y = if (fitsBelow) below else (hole.top - gap - cardHeight)
                    IntOffset(0, y.coerceAtLeast(margin).roundToInt())
                }
                .onSizeChanged { cardHeight = it.height }
                .background(Ink.Card, CardShape)
                .padding(18.dp),
        ) {
            Column(Modifier.fillMaxWidth()) {
                Text("${index + 1} / $total", style = Type.LabelSmall, color = Ink.Mute)
                Spacer(Modifier.height(6.dp))
                Text(title, style = Type.Section, color = Ink.Text)
                Spacer(Modifier.height(6.dp))
                Text(body, style = Type.Caption, color = Ink.Body)
                Spacer(Modifier.height(14.dp))
                Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                    // The way out is only offered while there is something left
                    // to skip.
                    if (!last) {
                        LinkText(stringResource(R.string.action_skip_tour), onSkip, style = Type.LinkSmall)
                    }
                    Spacer(Modifier.weight(1f))
                    PrimaryButton(
                        label = stringResource(if (last) R.string.action_got_it else R.string.action_next),
                        onClick = onNext,
                        height = 40.dp,
                        style = Type.ButtonSmall,
                    )
                }
            }
        }
    }
}
